common/rocksdb_env_s3.cpp

```diff
@@ -0,0 +1,432 @@
+///
+/// http://www.apache.org/licenses/LICENSE-2.0
+
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+//
+// @author ghan (ghan@pinterest.com)
+//
+
+#include "common/rocksdb_env_s3.h"
+
+#include <string>
+
+#include "common/rocksdb_glogger/rocksdb_glogger.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "rocksdb/status.h"
+
+DECLARE_bool(s3_direct_io);
+
+namespace rocksdb {
+
+// open a file for sequential reading
+Status S3Env::NewSequentialFile(const std::string& fname,
+                                std::unique_ptr<SequentialFile>* result,
+                                const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
+  auto local_full_path = local_directory_ + fname;
+  auto st = posix_env_->NewSequentialFile(local_full_path, result, options);
+
+  if (!st.ok()) {
+    // If file doesnt exist in local, we copy the file to the local storage from S3
+    auto resp = s3_util_->getObject(s3_key_prefix_ + fname, local_full_path, FLAGS_s3_direct_io);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when copying the file from S3 to local: "
+                 << resp.Error();
+      return Status::IOError();
+    }
+  }
+
+  // we successfully copied the file, try opening it locally now
+  st = posix_env_->NewSequentialFile(local_full_path, result, options);
+
+  return st;
+}
+
+// open a file for random reading
+Status S3Env::NewRandomAccessFile(const std::string& fname,
+                                  std::unique_ptr<RandomAccessFile>* result,
+                                  const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  // We read first from local storage and then from S3.
+  auto local_full_path = local_directory_ + fname;
+  auto st = posix_env_->NewRandomAccessFile(local_full_path, result, options);
+
+  if (!st.ok()) {
+    // If file doesnt exist in local, we copy the file to the local storage from S3
+    auto resp = s3_util_->getObject(s3_key_prefix_ + fname, local_full_path, FLAGS_s3_direct_io);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when copying the file from S3 to local: "
+                 << resp.Error();
+      return Status::IOError();
+    }
+  }
+
+  // we successfully copied the file, try opening it locally now
+  st = posix_env_->NewRandomAccessFile(local_full_path, result, options);
+
+  return st;
+}
+
+// S3WritableFile is a wrapper for writing a file in S3. We will perform the
+// operations in a local tmp file and upload to S3 when closing the file.
+class S3WritableFile : public WritableFile {
+ private:
```

Comment:
Convention is to put the `public` section first, followed by `protected` and then finally `private`

Reply:
good catch!